### PR TITLE
Public API for Pins

### PR DESCRIPTION
**Pull request by yschimke**
Addressing https://github.com/square/okhttp/issues/4049

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt [hunk 1]
```diff
@@ -212,16 +213,7 @@ class CertificatePinner internal constructor(
...
-    var result: List<Pin> = emptyList()
-    for (pin in pins) {
-      if (pin.matches(hostname)) {
-        if (result.isEmpty()) result = mutableListOf()
-        (result as MutableList<Pin>).add(pin)
-      }
-    }
-    return result
-  }
+  fun findMatchingPins(hostname: String): List<Pin> = pins.filterList { it.matches(hostname) }
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/CertificatePinner.kt:216]: OMG Kotlin nice

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt [hunk 2]
```diff
@@ -247,9 +239,9 @@ class CertificatePinner internal constructor(
     return result
   }
 
-  internal data class Pin(
+  data class Pin internal constructor(
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/CertificatePinner.kt:242]: internal constructors on a data class are almost useless thanks to .copy().

Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/CertificatePinner.kt:242]: ... wanna remove the data class bit here and make this a plain class?

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt [hunk 3]
```diff
@@ -276,6 +268,31 @@ class CertificatePinner internal constructor(
     }
 
     override fun toString(): String = hashAlgorithm + hash.base64()
+
+    companion object {
+      @JvmStatic
+      fun newPin(pattern: String, pin: String): Pin {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/CertificatePinner.kt:274]: make this into the class constructor? If it’s not a data class it’s easy to write that way

File: okhttp/src/main/java/okhttp3/internal/Util.kt [hunk 4]
```diff
@@ -354,7 +390,10 @@ fun Source.skipAll(duration: Int, timeUnit: TimeUnit): Boolean {
  * source is helpful, such as when doing so completes a cache body or frees a socket connection for
  * reuse.
  */
-fun Source.discard(timeout: Int, timeUnit: TimeUnit): Boolean = try {
+fun Source.discard(
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/Util.kt:393]: what’s with these changes?

File: okhttp/src/main/java/okhttp3/internal/Util.kt [hunk 5]
```diff
@@ -589,3 +636,14 @@ fun Exception.withSuppressed(suppressed: List<Exception>): Throwable = apply {
 
   for (e in suppressed) addSuppressed(e)
 }
+
+inline fun <T> Iterable<T>.filterList(predicate: (T) -> Boolean): List<T> {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/Util.kt:640]: dig

File: okhttp/src/main/java/okhttp3/internal/Util.kt (hunk 4 above)
Comment (yschimke) [reply okhttp/src/main/java/okhttp3/internal/Util.kt:393]: I'll revert.  I think came from the Square code formatter in Intellij?

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt [hunk 6]
```diff
@@ -275,7 +295,7 @@ class CertificatePinner internal constructor(
       }
     }
 
-    override fun toString(): String = hashAlgorithm + hash.base64()
+    override fun toString(): String = hashAlgorithm + "/" + hash.base64()
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/CertificatePinner.kt:298]: This seems like the better external api, with the slash removed

**Comment by yschimke** [comment]
Any thoughts on API for using a single Pin?  The loop in CertificatePinner is good to cache the tested certificate hash, but as an API we probably want some boolean test method?

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt [hunk 7]
```diff
@@ -247,14 +239,42 @@ class CertificatePinner internal constructor(
     return result
   }
 
-  internal data class Pin(
+  class Pin internal constructor(
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/CertificatePinner.kt:242]: Public constructor?

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt (hunk 6 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/CertificatePinner.kt:298]: Definitely.

Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/CertificatePinner.kt:298]: Template string?

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt [hunk 8]
```diff
@@ -317,29 +357,7 @@ class CertificatePinner internal constructor(
     internal fun X509Certificate.toSha1ByteString(): ByteString =
         publicKey.encoded.toByteString().sha1()
 
-    internal fun X509Certificate.toSha256ByteString(): ByteString =
+    fun X509Certificate.toSha256ByteString(): ByteString =
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/CertificatePinner.kt:360]: Rename to sha256Hash and make the other one public as sha1Hash? That's as close as we can get to the Pin API.

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt [hunk 9]
```diff
@@ -165,11 +166,11 @@ class CertificatePinner internal constructor(
 
       for (pin in pins) {
         when (pin.hashAlgorithm) {
-          "sha256/" -> {
+          "sha256" -> {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/CertificatePinner.kt:169]: I think the body of this for loop to be a method, `Pin.matchesCertificate(certificates: List<Certificate>)`.

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt [hunk 10]
```diff
@@ -247,14 +239,42 @@ class CertificatePinner internal constructor(
...
+        }
+        pin.startsWith("sha256/") -> {
+          this.hashAlgorithm = "sha256"
+          this.hash = pin.substring("sha256/".length).decodeBase64() ?: throw IllegalArgumentException("Invalid pin hash: $pin")
+        }
+        else -> throw IllegalArgumentException("pins must start with 'sha256/' or 'sha1/': $pin")
+      }
+    }
+
     fun matches(hostname: String): Boolean {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/CertificatePinner.kt:278]: Rename to matchesHostname so we have matchesHostname and matchesCertificates?

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt [hunk 11]
```diff
@@ -275,12 +293,38 @@ class CertificatePinner internal constructor(
...
+      result = 31 * result + hashAlgorithm.hashCode()
+      result = 31 * result + hash.hashCode()
+      return result
+    }
   }
 
   /** Builds a configured certificate pinner. */
   class Builder {
-    private val pins = mutableListOf<Pin>()
+    val pins = mutableListOf<Pin>()
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/CertificatePinner.kt:327]: Added to meet original request

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt [hunk 12]
```diff
@@ -133,8 +134,8 @@ import okio.ByteString.Companion.toByteString
  */
 @Suppress("NAME_SHADOWING")
 class CertificatePinner internal constructor(
-  private val pins: Set<Pin>,
-  internal val certificateChainCleaner: CertificateChainCleaner?
+  val pins: Set<Pin>,
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/CertificatePinner.kt:137]: Added to meet original request

**Pull request merged**